core/trino-main/src/main/java/io/trino/sql/gen/InCodeGenerator.java

```diff
@@ -91,10 +92,9 @@ public InCodeGenerator(SpecialForm specialForm)
     @VisibleForTesting
     static SwitchGenerationCase checkSwitchGenerationCase(Type type, List<RowExpression> values)
     {
-        if (values.size() > 32) {
-            // 32 is chosen because
-            // * SET_CONTAINS performs worst when smaller than but close to power of 2
-            // * Benchmark shows performance of SET_CONTAINS is better at 50, but similar at 25.
+        if (values.size() >= 10 || (type instanceof VarcharType && values.size() >= 5)) {
```

Comment:
What about other character types or types registered by plugins?
Could it be generalized maybe (e.g java type is `Slice`)?
Maybe it should just be `5`?

What happens when element is not in set, do we fallback to `defaultLabel` which is expensive to evaluate?

Reply:
> What about other character types or types registered by plugins?
> Could it be generalized maybe (e.g java type is `Slice`)?
> Maybe it should just be `5`?

It could be - it would be a bit slower (at least the current limited benchmarks show that). But logic would be simpler.

> 
> What happens when element is not in set, do we fallback to `defaultLabel` which is expensive to evaluate (it looks so for `DIRECT_SWITCH`)?

It looks so. Why do you think it is a problem. I do not see an alterantive.

Reply:
> It looks so. Why do you think it is a problem. I do not see an alterantive.

Why for `DIRECT_SWITCH` we cannot return `noMatch` instead of going to `defaultLabel`?
Similarly for `SET_CONTAINS`. It looks like we get penalized by using `defaultLabel` every time we miss match in simpler case.

Reply:
If i understand the code correctly the `defaultCaseBlock` is used for matching against non-constant expressions (and maybe some other cases too, not sure).

So if you have statement:
`x in 1, 5, some_column`

then matching against `1` and `5` will be handled by `switchBlock` and matching against `some_column` will be handled by `defaultCaseBlock`